src/Generating/EnderDragonFightStructuresGen.cpp

```diff
@@ -0,0 +1,191 @@
+{
+	// If not in the chunks to write
+	if ((a_ChunkDesc.GetChunkX() > 6) ||
+		(a_ChunkDesc.GetChunkX() < -6) ||
+		(a_ChunkDesc.GetChunkZ() > 6) ||
+		(a_ChunkDesc.GetChunkZ() < -6))
+	{
+		return;
+	}
+	// Places the exit portal
+	auto Coords = a_ChunkDesc.GetChunkCoords();
+	if (Coords == cChunkCoords({0, 0}))
+	{
+		auto EnderDragon = std::make_unique<cEnderDragon>();
+		EnderDragon->SetPosition({0.0, 80.0, 0.0});
+		a_ChunkDesc.GetEntities().emplace_back(std::move(EnderDragon));
+		a_ChunkDesc.WriteBlockArea(m_Fountain, static_cast<int>(FloorC(-m_Fountain.GetSizeX() / 2)), 62, static_cast<int>(FloorC(-m_Fountain.GetSizeX() / 2)), cBlockArea::msSpongePrint);
+		return;
+	}
+	else if (Coords == cChunkCoords({-1, 0}))
+	{
+		a_ChunkDesc.WriteBlockArea(m_Fountain, cChunkDef::Width - static_cast<int>(FloorC(m_Fountain.GetSizeX() / 2)), 62, static_cast<int>(FloorC(-m_Fountain.GetSizeZ() / 2)), cBlockArea::msSpongePrint);
+		return;
+	}
+	else if (Coords == cChunkCoords({0, -1}))
+	{
+		a_ChunkDesc.WriteBlockArea(m_Fountain, static_cast<int>(FloorC(-m_Fountain.GetSizeX() / 2)), 62, cChunkDef::Width - static_cast<int>(FloorC(m_Fountain.GetSizeZ() / 2)), cBlockArea::msSpongePrint);
+		return;
+	}
+	else if (Coords == cChunkCoords({-1, -1}))
+	{
+		a_ChunkDesc.WriteBlockArea(m_Fountain, cChunkDef::Width - static_cast<int>(FloorC(m_Fountain.GetSizeX() / 2)), 62, cChunkDef::Width - static_cast<int>(FloorC(m_Fountain.GetSizeZ() / 2)), cBlockArea::msSpongePrint);
+		return;
+	}
+	auto It = m_TowerPos.find(Coords);
+	if (It == m_TowerPos.end())
+	{
+		return;
+	}
+	PlaceTower(a_ChunkDesc, It->second);
+}
+
+
+
+
+
+void cEnderDragonFightStructuresGen::PlaceTower(cChunkDesc &a_ChunkDesc, const Vector3i & a_AbsPos)
+{
+	auto Pos = cChunk::AbsoluteToRelative(a_AbsPos);
+	sTowerProperties Properties = {0, 0, false};
+	// Choose random height
+	unsigned long Index = static_cast<unsigned long>(m_Noise.IntNoise3DInt(a_AbsPos)) % m_TowerProperties.size();
+	do
+	{
+		Index = (Index + 1) % m_TowerProperties.size();
+		Properties = m_TowerProperties[Index];
+	} while (Properties.m_Height == 0);
+
+	// Make sure the tower does not cross chunk boarders
+	Pos.x = Clamp(Pos.x, Properties.m_Radius - 1, cChunkDef::Width - Properties.m_Radius);
+	Pos.z = Clamp(Pos.z, Properties.m_Radius - 1, cChunkDef::Width - Properties.m_Radius);
+
+	// Place obsidian pillar
+	for (int Y = 0; Y <= Properties.m_Height - 2; Y++)
+	{
+		for (int X = -Properties.m_Radius; X < Properties.m_Radius; X++)
+		{
+			for (int Z = -Properties.m_Radius; Z < Properties.m_Radius; Z++)
+			{
+				Vector3i NewPos = {Pos.x + X, Y, Pos.z + Z};
+				// The 3 was achieved by trial and error till the shape matched the notchian implementation
+				if ((NewPos - Vector3i(Pos.x, Y, Pos.z)).SqrLength() < Properties.m_Radius * Properties.m_Radius - 3)
+				{
+					a_ChunkDesc.SetBlockType(NewPos.x, NewPos.y, NewPos.z, E_BLOCK_OBSIDIAN);
+				}
+			}
+		}
+	}
```

Comment:
The if-condition does not use `Y` at all here, it would be much more efficient if the Y-loop was the inner loop here.